### ISSUE TITLE
OAuthPrompt not working within Teams

### ISSUE THREAD

**Issue by WueF** [microsoft/botbuilder-dotnet#6071]
## Version
BotFramework packages: 4.15.0
.NET 6.0

Tested on:
Teams client (Microsoft Teams  1.4.00.32771 (64 bits))
Google Chrome (96.0.4664.45)

## Describe the bug
After going to OAuth prompt in Teams, login card is displayed properly. When you press "Sign In" button, error "This action can't be performed since the app does not exist or has been uninstalled" appears under card. No network movement is recorded.

Behaviour is the same in Teams App and in Teams in browser.

## To Reproduce
1. Create bot with Teams channel, and OAuthPrompt being used
2. Go to OAuthPrompt  dialog in Teams
2. Click "Sign In" button
3. See error

## Expected behavior
OAuth login should work as expected, or at least all fallback scenarios like magic code should work

## Screenshots
![Teamsbug](https://user-images.githubusercontent.com/10574786/146970419-61baa64f-1995-4799-b107-70254b218183.jpg)

## Additional context
1. Tested on few environments
2. Bot is being used via Teams channel without App package (Test instance)
3. Created additional OAuth config in Bot channel configuration, still the same issue
4. Behaviour is same in Teams App (Windows) and in Teams Web (browser)
5. In browser, no network movement has been noticed after clicking "Sign In" button while using debug tools in browser

**Comment by anishprasad01** [comment]
Hi @WueF,

There is already an [open issue](https://github.com/microsoft/BotFramework-Composer/issues/8843) (it does not appear to be specific to Composer, but rather to Teams) under investigation regarding this problem. 

I am closing this issue in favor of the existing one. Please contribute there instead.

**Comment by WueF** [comment]
@anishprasad01 Please re-open this ticket.
I've never asked for this one to be closed.
The other one, which you we're refering to, was closed by original author. Our bot is still affected, so please continue investigation.
@darrenj Can you please take a look on that?

**Comment by WueF** [comment]
![image](https://user-images.githubusercontent.com/10574786/147715393-8b2c0b9d-ca05-4a7f-91be-46e51bcd433a.png)
As for today, issue is still valid on our side, please open this ticket.

**Comment by anishprasad01** [comment]
Have your tried creating an app package and adding that to teams, as well as making sure `token.botframework.com` is in your supported domains? Talking to the bot directly by id can sometimes cause issues, so we want to make sure that this isn't a factor.

**Comment by WueF** [comment]
@anishprasad01 
1. We have app-package that runs production resource (now with older codebase, .NET Core 3.1/BotFramework 4.11), and in package federation login works fine
2. I'm not able to test the bot now to promote it to production...
3. Adding package within same teams tenant is impossible (too many corporate rules), I might do a check on other tenant with package, but that doesn't solve my issue, that now other users won't be able to access resources by ID, which apparently is pretty useful thing.

Are you saying, that probability of this issue is lower when accessing via package/it doesn't happen then?

Because I won't risk 120k users to see "sorry" message.

**Comment by anishprasad01** [comment]
All I'm saying is that sometime auth can have issues when accessing the bot by id. That said, I don't yet know what's causing this issue, so I wouldn't assume switching to app packages will fix it.

**Comment by anishprasad01** [comment]
Just letting you know this issue hasn't been forgotten and we are still investigating.

**Comment by anishprasad01** [comment]
Not yet, unfortunately. I'm engaging people from the Teams side to try to get to the bottom of this.

**Comment by clearab** [comment]
Authentication will not work in Teams if the app is not installed via a Teams app package. If you talk to the bot by searching for its ID w/o installing, authentication will fail.

**Comment by Wajeed-msft** [comment]
@WueF - as @clearab  mentioned,  Testing your bot without installing it as an app can be useful to ensure your bot is available and responding, however it won't allow you to test the full breadth of Microsoft Teams functionality you may have added to your bot.

Ref: [Talk to your bot directly by ID](https://docs.microsoft.com/en-us/microsoftteams/platform/resources/bot-v3/bots-test#talk-to-your-bot-directly-by-id)
<img width="473" alt="image" src="https://user-images.githubusercontent.com/31851992/150727076-f6e9c0e7-fe55-4eba-a4f6-8410c01b0f34.png">

**Comment by WueF** [comment]
@Wajeed-msft For me it's regression from Bot Framework/Teams side. I am and I was fully aware what are limitations of using bot by ID and as mentioned in ticket, and in further discussion. And I'm currently awaiting for a month or so with releasing my new functionalities to production, cos my test bot on proper tenant cannot be tested properly, because authentication is not working. 

So according to your answer, I should promote bot to production with YOLO method, without possiblility to do proper integration tests, because "I may not be allowed to test the full breath"? 

Yes, I'm aware that cards/tabs/attachments aren't available within talking-by-id, but as I mentioned, in earlier versions of BF, we were sucessfully able to perform this scenario, and now we're not. For me it's a huge regression.

Also in mentioned doc, there's no word saying that authentication won't work. Moreover, your own code samples that we're working, are not now. So I'm pretty convinced, this is a bug and regression.

**Comment by anishprasad01** [comment]
@WueF 

I've been trying to repro this issue once again in conjunction with folks from the Teams side. I apologize about retreading ground here, I'd just like to reconfirm a few things to make sure I fully understand the situation.

- You are using test resources within the same tenant as the production bot.
- You are contacting the bot by ID only for testing and attempting to sign-in, which we have established will not work.
- As I understand it, you would be using a Teams App Manifest in production (which should work), but you cannot install a test manifest due to corporate rules.

>  in earlier versions of BF, we were sucessfully able to perform this scenario, and now we're not.

By this you mean signing in with a bot that you're talking to by ID?

**Comment by WueF** [comment]
@anishprasad01 

1. Yes, same tenant.
2. Yes. And indeed, I've created package, tested it, and issue applies only when talking to bot by id
3. Yes, exactly. And that is more of business inconvenience. At this point I'm not able to show pre-production bots to business users. 

Of course for me, myself and I can create test tenant and perform everything on totally mocked ground, but there are users who aren't capable of creating their own test tenants etc.

And yes, I mean, we were somehow able to perform authentication in bot we were talking to by ID. We have identified issue, when I was demoing new features to new users (old ones had active sessions, so refresh token was working fine), but new users have seen message as shown. 


I understand, there are technical limitations etc, but as BF has generally this flow, that magic-code is fallback (it happens in emulator in example), why Teams channel behaves in such strange way, with strange and non-meaningful error message, and it's not clear from documenation, that OAuthPrompt (which is pretty generic) won't work in Teams channel without app-manifest, and there won't be any fallback.

I'm sure that, when I was first creating authentication in a bot, which is around 1,5 year ago, there were no problems like that. I was doing a lot of tests at different bots, and OAuthPrompt was somehow working. I just can't remember, whether while talking to bot by ID, there was proper browser-based-workflow, or magic-code fallback mechanism, but surly it was possible, and didn't cause blocking error.

**Comment by anishprasad01** [comment]
From what I'm hearing from the Teams side, due to security reasons, the OAuth by ID scenario was never intended behavior and was meant to work. How did you manage to get it working in the past?

Additionally, I appreciate you bringing the deficiency in the documentation to light. I will see about updating them so there is no further confusion.

**Comment by WueF** [comment]
@anishprasad01 I said, I can't remember which flow it went. But what I'm sure, that authenticating was possible by some mean.
Especially, as I have 12 independent users authenticated in test instance, which never had app package released.

Indeed, there's now a small note in attached article. 

Ok, so your final statement is, that OAuth flow will never work again in Teams without App package?
Also if something like that happens in future, would be nice to have it added to breaking changes of BF libs ;)

So to sum up:
1. OAuth flow was some time ago working without problems without AppPackage for Teams
2. Now it doesn't. It also leaves totally unmeaningful error message, so that each and every developer seeing this message will open a ticket here
3. With App package it works as expected, so nobody will care about it anyway
4. Now I can peacefully go climb a tree

I hope I didn't miss anything.

Closing.

**Comment by anishprasad01** [comment]
I can pass on your feedback to the Teams team, but yes, that is their final word. I can also see about getting the docs updated to be clear about this error message. Other than that, no, there's nothing else we can do.
